### PR TITLE
Kotlin conversion: Rest of the Funnel classes

### PR DESCRIPTION
**Pull request by sharvaniharan**
Wrapped up the Funnel package before starting Butterknife usage.

### CODE REVIEW

**Comment by sharvaniharan** [comment]
Ready for review @cooltey @dbrant

File: app/src/main/java/org/wikipedia/analytics/ShareAFactFunnel.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    private val pageTitle: String = pageTitle.displayText
+
+    override fun preprocessData(eventData: JSONObject): JSONObject? {
+        preprocessData(eventData, "tutorial_feature_enabled", true)
+        preprocessData(eventData, "tutorial_shown", 0)
+        return super.preprocessData(eventData)
+    }
+
+    private fun logAction(action: String, text: String, shareMode: ShareMode? = null) {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/ShareAFactFunnel.kt:18]: Should we make the `text: String` nullable?

File: app/src/main/java/org/wikipedia/analytics/ShareAFactFunnel.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        var textStr: String? = text
+        if (textStr != null) {
+            textStr = textStr.substring(0, MAX_LENGTH.coerceAtMost(textStr.length))
+        }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/ShareAFactFunnel.kt:22]: If the `text` is nullable, would this be better?
```
        val textStr = text.orEmpty().substring(0, MAX_LENGTH.coerceAtMost(text.orEmpty().length))

```

File: app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+import java.util.*
+
+class SuggestedEditsFunnel private constructor(app: WikipediaApp, private val invokeSource: InvokeSource) :
+        TimedFunnel(app, SCHEMA_NAME, REV_ID, SAMPLE_LOG_ALL) {
+
+    private val parentSessionToken: String?
+    private var helpOpenedCount = 0
+    private var contributionsOpenedCount = 0
+    private val statsCollection = SuggestedEditStatsCollection()
+    private val uniqueTitles: MutableList<String> = ArrayList()
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt:23]: Could be `uniqueTitles = mutableListOf<String>()`

File: app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            }
+            action === DescriptionEditActivity.Action.ADD_CAPTION -> {
+                statsCollection.addCaptionStats
+            }
+            action === DescriptionEditActivity.Action.TRANSLATE_CAPTION -> {
+                statsCollection.translateCaptionStats
+            }
+            else -> {
+                return
+            }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt:69]: Can be `else -> return`

File: app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        @SerializedName("cxl")
+        var cancels = 0
+
+        @SerializedName("suc")
+        var successes = 0
+
+        @SerializedName("fl")
+        var failures = 0
+        val isEmpty: Boolean
+            get() = impressions == 0 && clicks == 0 && suggestionsClicked == 0 && cancels == 0 && successes == 0 && failures == 0
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt:183]: Can be `val isEmpty = impressions ...`

File: app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    companion object {
+        private var INSTANCE: SuggestedEditsFunnel? = null
+        private const val SCHEMA_NAME = "MobileWikiAppSuggestedEdits"
+        private const val REV_ID = 18949003
+        private const val SUGGESTED_EDITS_UI_VERSION = "1.0"
+        private const val SUGGESTED_EDITS_API_VERSION = "1.0"
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt:191]: Please have a proper order.

File: app/src/main/java/org/wikipedia/analytics/TabFunnel.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    fun logSelect(size: Int, index: Int) {
+        log("select", size, index)
+    }
+
+    fun logCancel(size: Int) {
+        log("cancel", size)
+    }
+
+    private fun log(action: String, size: Int, index: Int? = null) {
+        val params: MutableList<Any> = ArrayList()
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/TabFunnel.kt:33]: Can be `val params = mutableListOf<Any>()`

File: app/src/main/java/org/wikipedia/analytics/TimedFunnel.kt [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.concurrent.TimeUnit
+
+abstract class TimedFunnel @JvmOverloads constructor(app: WikipediaApp, schemaName: String, revision: Int, sampleRate: Int, wiki: WikiSite? = null) :
+        Funnel(app, schemaName, revision, sampleRate, wiki) {
+
+    private var startTime: Long
+    private var pauseTime: Long = 0
+
+    init {
+        startTime = System.currentTimeMillis()
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/TimedFunnel.kt:15]: Can this be initialized at line `11`?

File: app/src/main/java/org/wikipedia/analytics/TimedFunnel.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    /** Override me for deviant implementations.  */
+    private val durationFieldName: String
+        get() = "time_spent"
+
+    private val duration: Long
+        get() = System.currentTimeMillis() - startTime
+
+    private val durationSeconds: Long
+        get() = TimeUnit.MILLISECONDS.toSeconds(duration)
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/TimedFunnel.kt:26]: These can be simplified without having a `get()`.

File: app/src/main/java/org/wikipedia/analytics/UserContributionFunnel.kt [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    fun logIpBlock() {
+        log("action", "ip_block")
+    }
+
+    companion object {
+        private var INSTANCE: UserContributionFunnel? = null
+        private const val SCHEMA_NAME = "MobileWikiAppUserContribution"
+        private const val REV_ID = 20217330
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/UserContributionFunnel.kt:74]: Please have a proper order.

File: app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+                INSTANCE = SuggestedEditsFunnel(WikipediaApp.getInstance(), invokeSource)
+            }
+            return INSTANCE
+        }
+
+        fun get(): SuggestedEditsFunnel? {
+            return if (INSTANCE != null && INSTANCE!!.invokeSource != InvokeSource.SUGGESTED_EDITS) {
+                INSTANCE
+            } else Companion[InvokeSource.SUGGESTED_EDITS]
+        }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt:211]: Looks like these static functions should return non-null variable.

These can be done by adding `!!` or `as SuggestedEditsFunnel`

File: app/src/main/java/org/wikipedia/analytics/UserContributionFunnel.kt [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+        fun get(): UserContributionFunnel? {
+            if (INSTANCE == null) {
+                INSTANCE = UserContributionFunnel()
+            }
+            return INSTANCE
+        }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/analytics/UserContributionFunnel.kt:81]: Looks like this static function should return non-null variable and can be done by adding `!!` or `as UserContributionFunnel`.

**Review by cooltey — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/analytics/AppearanceChangeFunnel.kt [hunk 13]
```diff
@@ -36,7 +36,7 @@ class AppearanceChangeFunnel(app: WikipediaApp, wiki: WikiSite, private val sour
 
     override fun preprocessData(eventData: JSONObject): JSONObject {
         preprocessData(eventData, "invoke_source", source.ordinal)
-        return super.preprocessData(eventData)
+        return super.preprocessData(eventData)!!
```
[the remark this answers is not in the document]
Comment (sharvaniharan) [reply app/src/main/java/org/wikipedia/analytics/AppearanceChangeFunnel.kt:39]: When we get a non-null eventdata we will definitely send back a non-null JSONObject @cooltey So will change the parent.

**Pull request merged**